
* estimating read length of fastq

  awk 'NR%4 == 2 {lengths[length($0)]++} END {for (l in lengths) {print l, lengths[l]}}' file.fastq

  git@github.com:maasha/biopieces.git


* 2016-01-04 Adam:
    Are you able to calculate the percent spliced in (PSI) for each exon - see
    Dmitri's paper on this here (http://www.ncbi.nlm.nih.gov/pubmed/23172860)
    - from the data you have. I think in the absence of full transcript level
    quantification this would be a very useful property to capture - e.g. an
    exon may be well-supported in a highly expressed gene but may be included
    in relatively few transcripts.


* 2018-03-15 rnaseqer API:
   https://www.ebi.ac.uk/fg/rnaseq/api/
   https://github.com/nunofonseca/irap

https://www.ebi.ac.uk/fg/rnaseq/api/tsv/0/getRunsByOrganism/homo_sapiens/

* 2019-04-16
** How to store TSL evidence
- tabidx PSL files per experiment
- database table combined or per experiment
** decision:
- tabidx per PSL
bgzip  foo.psl
tabix --sequence 14 --begin 16 --end 17 --zero-based foo.psl.gz
- store results in table



* output:

GenbankProblemCases(self.conn)

* tsl reorg
- evidenceMeta
- evidenceAligns
- supportEval

* database structure
EvidenceSource
EvidenceProcess can be chained
Results  - file
sort-ed tsv files vs database

